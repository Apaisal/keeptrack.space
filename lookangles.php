<?php
header('Content-Type: text/html; charset=utf-8');
?>
<!doctype html>
<html>
  <head>
    <script src="jquery-2.2.2.min.js"></script>
    <script src="satellite.min.js"></script>
  </head>
  <body>
    <table id="looks"  cellpadding="10" style="border: 1px; padding: 10px;">
    </table>
    </div>
    <script src="lookangles.js"></script>
  </body>
</html>
