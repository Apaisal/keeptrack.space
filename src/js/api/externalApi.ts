interface keepTrackApiInterface {
  register: (params: { method: string; cbName: string; cb: () => void }) => void;
  unregister: (params: { method: string; cbName: string }) => void;
  callbacks: any;
  methods: any;
  programs: any;
}

const keepTrackApi: keepTrackApiInterface = {
  register: (params: { method: string; cbName: string; cb: () => void }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      // Add the callback
      keepTrackApi.callbacks[params.method].push({ name: params.cbName, cb: params.cb });
    } else {
      throw new Error(`Invalid callback "${params.method}"!`);
    }
    return;
  },
  unregister: (params: { method: string; cbName: string }) => {
    // If this is a valid callback
    if (typeof keepTrackApi.callbacks[params.method] !== 'undefined') {
      for (let i = 0; i < keepTrackApi.callbacks[params.method].length; i++) {
        if (keepTrackApi.callbacks[params.method][i].name == params.cbName) {
          keepTrackApi.callbacks[params.method].splice(i, 1);
          return;
        }
      }
      // If we got this far, it means we couldn't find the callback
      throw new Error(`Callback "${params.cbName} not found"!`);
    } else {
      // Couldn't find the method
      throw new Error(`Invalid callback "${params.method}"!`);
    }
  },
  callbacks: {
    selectSatData: [],
    updateSelectBox: [],
  },
  methods: {
    selectSatData: (sat: any, satId: number) => {
      for (let i = 0; i < keepTrackApi.callbacks.selectSatData.length; i++) {
        keepTrackApi.callbacks.selectSatData[i].cb(sat, satId);
      }
    },
    updateSelectBox: (sat: any) => {
      for (let i = 0; i < keepTrackApi.callbacks.updateSelectBox.length; i++) {
        keepTrackApi.callbacks.updateSelectBox[i].cb(sat);
      }
    },
  },
  programs: {
    timeManager: {},
    settingsManager: {},
    ColorScheme: {},
    drawManager: {},
    missileManager: {},
    objectManager: {},
    orbitManager: {},
    photoManager: {},
    satSet: {},
    satellite: {},
    searchBox: {},
    sensorManager: {},
    starManager: {},
    sMM: {},
    uiManager: {},
    uiInput: {},
  },
};

(<any>window).keepTrackApi = keepTrackApi;
export { keepTrackApi };
