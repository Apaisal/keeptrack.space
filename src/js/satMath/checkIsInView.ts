import { SensorObject } from '../api/keepTrackTypes';

// TODO: Future Idea
/*
// satellite.createManeuverAnalyst = (satId, incVariation, meanmoVariation, rascVariation) => {
//   const { timeManager, satSet } = keepTrackApi.programs;
//   // TODO This needs rewrote from scratch to bypass the satcruncher
//   var mainsat = satSet.getSat(satId);
//   var origsat = mainsat;
//   // Launch Points are the Satellites Current Location
//   var TEARR = mainsat.getTEARR();
//   var launchLat, launchLon, alt;
//   launchLat = satellite.degreesLat(TEARR.lat);
//   launchLon = satellite.degreesLong(TEARR.lon);
//   alt = TEARR.alt;
//   var upOrDown = mainsat.getDirection();
//   var currentEpoch = satellite.currentEpoch(timeManager.simulationTimeObj);
//   mainsat.TLE1 = mainsat.TLE1.substr(0, 18) + currentEpoch[0] + currentEpoch[1] + mainsat.TLE1.substr(32);
//   var TLEs;
//   // Ignore argument of perigee for round orbits OPTIMIZE
//   if (mainsat.apogee - mainsat.perigee < 300) {
//     TLEs = satellite.getOrbitByLatLon(mainsat, launchLat, launchLon, upOrDown, timeManager.simulationTimeObj);
//   } else {
//     TLEs = satellite.getOrbitByLatLon(mainsat, launchLat, launchLon, upOrDown, timeManager.simulationTimeObj, alt);
//   }
//   var TLE1 = TLEs[0];
//   var TLE2 = TLEs[1];
//   //   var breakupSearchString = '';
//   satId = satSet.getIdFromObjNum(80000);
//   var sat = satSet.getSat(satId);
//   sat = origsat;
//   let iTLE1 = '1 ' + 80000 + TLE1.substr(7);
//   let iTLEs;
//   // Ignore argument of perigee for round orbits OPTIMIZE
//   if (sat.apogee - sat.perigee < 300) {
//     iTLEs = satellite.getOrbitByLatLon(sat, launchLat, launchLon, upOrDown, timeManager.simulationTimeObj, 0, rascVariation);
//   } else {
//     iTLEs = satellite.getOrbitByLatLon(sat, launchLat, launchLon, upOrDown, timeManager.simulationTimeObj, alt, rascVariation);
//   }
//   iTLE1 = iTLEs[0];
//   let iTLE2 = iTLEs[1];
//   // For the first 30
//   var inc = TLE2.substr(8, 8);
//   inc = (parseFloat(inc) + incVariation).toPrecision(7);
//   inc = inc.split('.');
//   inc[0] = inc[0].substr(-3, 3);
//   if (inc[1]) {
//     inc[1] = inc[1].substr(0, 4);
//   } else {
//     inc[1] = '0000';
//   }
//   inc = (inc[0] + '.' + inc[1]).toString();
//   inc = stringPad.padEmpty(inc, 8);
//   // For the second 30
//   var meanmo: any = iTLE2.substr(52, 10);
//   meanmo = (parseFloat(meanmo) * meanmoVariation).toPrecision(10);
//   // meanmo = parseFloat(meanmo - (0.005 / 10) + (0.01 * ((meanmoIterat + 1) / 10))).toPrecision(10);
//   meanmo = meanmo.split('.');
//   meanmo[0] = meanmo[0].substr(-2, 2);
//   if (meanmo[1]) {
//     meanmo[1] = meanmo[1].substr(0, 8);
//   } else {
//     meanmo[1] = '00000000';
//   }
//   meanmo = (meanmo[0] + '.' + meanmo[1]).toString();
//   iTLE2 = '2 ' + 80000 + ' ' + inc + ' ' + iTLE2.substr(17, 35) + meanmo + iTLE2.substr(63);
//   sat = satSet.getSat(satId);
//   sat.TLE1 = iTLE1;
//   sat.TLE2 = iTLE2;
//   sat.active = true;
//   if (satellite.altitudeCheck(iTLE1, iTLE2, timeManager.simulationTimeObj) > 1) {
//     satSet.satCruncher.postMessage({
//       type: 'satEdit',
//       id: satId,
//       TLE1: iTLE1,
//       TLE2: iTLE2,
//     });
//     // TODO: This belongs in main or uiManager
//     // orbitManager.updateOrbitBuffer(satId, true, iTLE1, iTLE2);
//   } else {
//     console.debug('Breakup Generator Failed');
//     return false;
//   }
//   // breakupSearchString += mainsat.sccNum + ',Analyst Sat';
//   // uiManager.doSearch(breakupSearchString);
//   return true;
// };
// satellite.findChangeOrbitToDock = (sat, sat2, propOffset, propLength) => {
//   const { satSet } = keepTrackApi.programs;
//   let closestInc = 0;
//   let closestRaan = 0;
//   let closestMeanMo = 1;
//   let minDistArray = {
//     dist: 1000000,
//   };
//   for (let incTemp = -1; incTemp <= 1; incTemp++) {
//     for (let raanTemp = -1; raanTemp <= 1; raanTemp++) {
//       for (let meanMoTemp = 0.95; meanMoTemp <= 1.05; meanMoTemp += 0.05) {
//         if (satellite.createManeuverAnalyst(sat.id, incTemp, meanMoTemp, raanTemp)) {
//           let minDistArrayTemp = satellite.findClosestApproachTime(satSet.getSatFromObjNum(80000), sat2, propOffset, propLength);
//           if (minDistArrayTemp.dist < minDistArray.dist) {
//             minDistArray = minDistArrayTemp;
//             // let closestInc = incTemp;
//             // let closestRaan = raanTemp;
//             // let closestMeanMo = meanMoTemp;
//             // console.log(`Distance: ${minDistArray.dist}`);
//             // console.log(`Time: ${minDistArray.time}`);
//             // console.log(satSet.getSatFromObjNum(80000));
//           }
//         }
//       }
//     }
//   }

//   console.log(`${sat.inclination + closestInc}`);
//   console.log(`${sat.raan + closestRaan}`);
//   console.log(`${sat.meanMotion * closestMeanMo}`);
//   satellite.createManeuverAnalyst(sat.id, closestInc, closestMeanMo, closestRaan);
// };
*/
export const checkIsInView = (sensor: SensorObject, rae: { rng: number; az: number; el: number }): boolean => {
  const { az, el, rng } = rae;

  if (sensor.obsminaz > sensor.obsmaxaz) {
    if (
      ((az >= sensor.obsminaz || az <= sensor.obsmaxaz) && el >= sensor.obsminel && el <= sensor.obsmaxel && rng <= sensor.obsmaxrange && rng >= sensor.obsminrange) ||
      ((az >= sensor.obsminaz2 || az <= sensor.obsmaxaz2) && el >= sensor.obsminel2 && el <= sensor.obsmaxel2 && rng <= sensor.obsmaxrange2 && rng >= sensor.obsminrange2)
    ) {
      return true;
    } else {
      return false;
    }
  } else {
    if (
      (az >= sensor.obsminaz && az <= sensor.obsmaxaz && el >= sensor.obsminel && el <= sensor.obsmaxel && rng <= sensor.obsmaxrange && rng >= sensor.obsminrange) ||
      (az >= sensor.obsminaz2 && az <= sensor.obsmaxaz2 && el >= sensor.obsminel2 && el <= sensor.obsmaxel2 && rng <= sensor.obsmaxrange2 && rng >= sensor.obsminrange2)
    ) {
      return true;
    } else {
      return false;
    }
  }
};
