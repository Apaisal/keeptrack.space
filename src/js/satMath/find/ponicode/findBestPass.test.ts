import { defaultSat, defaultSensor, keepTrackApiStubs } from '@app/js/api/apiMocks';
import { keepTrackApi } from '@app/js/api/keepTrackApi';
import { KeepTrackPrograms } from '@app/js/api/keepTrackTypes';
import * as findBestPass from '@app/js/satMath/find/findBestPass';
import * as apiMocks from '../../../api/apiMocks';
import * as SpaceObjectType from '../../../api/SpaceObjectType';
import * as checkIsInView from '../../lookangles/checkIsInView';

keepTrackApi.programs = <KeepTrackPrograms>(<unknown>{ ...keepTrackApi.programs, ...keepTrackApiStubs.programs });

// @ponicode
describe('findBestPass.findBestPass', () => {
  test('0', () => {
    let param2: any = [
      {
        static: true,
        linkIridium: true,
        linkGalileo: false,
        linkStarlink: true,
        obsmaxel2: 1,
        obsmaxrange2: 30,
        obsminrange2: 350,
        obsminel2: 70,
        obsmaxaz2: -100,
        obsminaz2: 100,
        alt: 50,
        beamwidth: 24,
        changeObjectInterval: -5.48,
        country: 'US',
        lat: -90,
        linkAehf: true,
        linkWgs: false,
        lon: -45,
        name: 'George',
        observerGd: { lat: 45, lon: 180, alt: 1 },
        obsmaxaz: -100,
        obsmaxel: 100,
        obsmaxrange: 100,
        obsminaz: 1,
        obsminel: -1.0,
        obsminrange: 1,
        shortName: 'www.google.com',
        staticNum: 10,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.PAYLOAD_OWNER,
        url: 'https://twitter.com/path?abc',
        volume: true,
        zoom: "['39.2233', '-78.8613']",
      },
      {
        static: false,
        linkIridium: true,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: -5.48,
        obsmaxrange2: 70,
        obsminrange2: 4,
        obsminel2: 320,
        obsmaxaz2: -100,
        obsminaz2: 1,
        alt: 1,
        beamwidth: 480,
        changeObjectInterval: -5.48,
        country: 'US',
        lat: 45,
        linkAehf: false,
        linkWgs: true,
        lon: 45,
        name: 'Jean-Philippe',
        observerGd: { lat: -45, lon: 90, alt: 1 },
        obsmaxaz: -5.48,
        obsmaxel: 1,
        obsmaxrange: 70,
        obsminaz: 0,
        obsminel: -1.0,
        obsminrange: 100,
        shortName: 'http://www.croplands.org/account/confirm?t=',
        staticNum: 10,
        sun: 'v1.2.4',
        type: SpaceObjectType.SpaceObjectType.PAYLOAD_OWNER,
        url: 'ponicode.com',
        volume: false,
        zoom: "['41.1994', '-115.9206']",
      },
    ];
    let result: any = findBestPass.findBestPass(apiMocks.defaultSat, param2);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let param2: any = [
      {
        static: true,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: false,
        obsmaxel2: 100,
        obsmaxrange2: 1,
        obsminrange2: 30,
        obsminel2: 70,
        obsmaxaz2: -5.48,
        obsminaz2: 100,
        alt: 5,
        beamwidth: 30,
        changeObjectInterval: -5.48,
        country: 'United States',
        lat: 45,
        linkAehf: true,
        linkWgs: true,
        lon: 180,
        name: 'Jean-Philippe',
        observerGd: { lat: 0, lon: 0, alt: 1 },
        obsmaxaz: 1,
        obsmaxel: -5.48,
        obsmaxrange: 4,
        obsminaz: 100,
        obsminel: 1.0,
        obsminrange: 1,
        shortName: 'http://base.com',
        staticNum: 1,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'Www.GooGle.com',
        volume: false,
        zoom: "['61.0865', '-144.7750']",
      },
      {
        static: true,
        linkIridium: false,
        linkGalileo: true,
        linkStarlink: false,
        obsmaxel2: 0,
        obsmaxrange2: 550,
        obsminrange2: 30,
        obsminel2: 50,
        obsmaxaz2: 1,
        obsminaz2: -100,
        alt: 5,
        beamwidth: 255,
        changeObjectInterval: 100,
        country: 'China',
        lat: 0,
        linkAehf: true,
        linkWgs: false,
        lon: 90,
        name: 'Michael',
        observerGd: { lat: -180, lon: 180, alt: 50 },
        obsmaxaz: 0,
        obsmaxel: -5.48,
        obsmaxrange: 70,
        obsminaz: 100,
        obsminel: 10.23,
        obsminrange: 1,
        shortName: 'https://api.telegram.org/',
        staticNum: 1000,
        sun: 'v1.2.4',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'https://twitter.com/path?abc',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
      {
        static: true,
        linkIridium: false,
        linkGalileo: true,
        linkStarlink: false,
        obsmaxel2: 1,
        obsmaxrange2: 520,
        obsminrange2: 100,
        obsminel2: 400,
        obsmaxaz2: -100,
        obsminaz2: -100,
        alt: 5,
        beamwidth: 544,
        changeObjectInterval: 100,
        country: 'France',
        lat: -45,
        linkAehf: false,
        linkWgs: false,
        lon: -180,
        name: 'Pierre Edouard',
        observerGd: { lat: 0, lon: -45, alt: 3 },
        obsmaxaz: 0,
        obsmaxel: 0,
        obsmaxrange: 100,
        obsminaz: -100,
        obsminel: -1.0,
        obsminrange: -100,
        shortName: 'https://twitter.com/path?abc',
        staticNum: 1000,
        sun: '1.0.0',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'www.google.com',
        volume: true,
        zoom: "['41.1994', '-115.9206']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: -5.48,
        obsmaxrange2: 320,
        obsminrange2: 50,
        obsminel2: 100,
        obsmaxaz2: 100,
        obsminaz2: 1,
        alt: 3,
        beamwidth: 390,
        changeObjectInterval: 1,
        country: 'GB',
        lat: -90,
        linkAehf: true,
        linkWgs: false,
        lon: -180,
        name: 'Michael',
        observerGd: { lat: 45, lon: 180, alt: 1 },
        obsmaxaz: 1,
        obsmaxel: 100,
        obsmaxrange: 30,
        obsminaz: -100,
        obsminel: 0.0,
        obsminrange: -100,
        shortName: 'Www.GooGle.com',
        staticNum: 10,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'ponicode.com',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: 0,
        obsmaxrange2: 30,
        obsminrange2: 410,
        obsminel2: 70,
        obsmaxaz2: 100,
        obsminaz2: 100,
        alt: 3,
        beamwidth: 544,
        changeObjectInterval: -5.48,
        country: 'GB',
        lat: 180,
        linkAehf: false,
        linkWgs: true,
        lon: -90,
        name: 'George',
        observerGd: { lat: 90, lon: 0, alt: 1 },
        obsmaxaz: 1,
        obsmaxel: 0,
        obsmaxrange: 380,
        obsminaz: 100,
        obsminel: -29.45,
        obsminrange: -100,
        shortName: 'http://www.croplands.org/account/confirm?t=',
        staticNum: 1000,
        sun: 'v4.0.0-rc.4',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'Www.GooGle.com',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
    ];
    let result: any = findBestPass.findBestPass(apiMocks.defaultSat, param2);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let param2: any = [
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: true,
        obsmaxel2: -5.48,
        obsmaxrange2: 50,
        obsminrange2: 400,
        obsminel2: 520,
        obsmaxaz2: 1,
        obsminaz2: -100,
        alt: 50,
        beamwidth: 16,
        changeObjectInterval: 1,
        country: 'FR',
        lat: -180,
        linkAehf: true,
        linkWgs: false,
        lon: 45,
        name: 'Michael',
        observerGd: { lat: -45, lon: -45, alt: 3 },
        obsmaxaz: -5.48,
        obsmaxel: -100,
        obsmaxrange: 70,
        obsminaz: -100,
        obsminel: -29.45,
        obsminrange: -5.48,
        shortName: 'https://',
        staticNum: 1,
        sun: '^5.0.0',
        type: SpaceObjectType.SpaceObjectType.RADAR_MEASUREMENT,
        url: 'ponicode.com',
        volume: true,
        zoom: "['-84.8577', '134.8560']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: true,
        obsmaxel2: 0,
        obsmaxrange2: 520,
        obsminrange2: 550,
        obsminel2: 350,
        obsmaxaz2: -100,
        obsminaz2: 100,
        alt: 5,
        beamwidth: 25,
        changeObjectInterval: 100,
        country: 'United States',
        lat: -180,
        linkAehf: true,
        linkWgs: true,
        lon: -90,
        name: 'Anas',
        observerGd: { lat: -90, lon: 0, alt: 1 },
        obsmaxaz: -100,
        obsmaxel: 1,
        obsmaxrange: 380,
        obsminaz: -100,
        obsminel: 0.0,
        obsminrange: 0,
        shortName: 'http://example.com/showcalendar.html?token=CKF50YzIHxCTKMAg',
        staticNum: 1000,
        sun: '^5.0.0',
        type: SpaceObjectType.SpaceObjectType.RADAR_MEASUREMENT,
        url: 'https://',
        volume: true,
        zoom: "['39.2233', '-78.8613']",
      },
    ];
    let result: any = findBestPass.findBestPass(apiMocks.defaultSat, param2);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let param2: any = [
      {
        static: false,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: true,
        obsmaxel2: -100,
        obsmaxrange2: 400,
        obsminrange2: 90,
        obsminel2: 4,
        obsmaxaz2: 0,
        obsminaz2: 100,
        alt: 3,
        beamwidth: 80.0,
        changeObjectInterval: -5.48,
        country: 'United States',
        lat: 0,
        linkAehf: true,
        linkWgs: false,
        lon: -180,
        name: 'Anas',
        observerGd: { lat: 45, lon: 180, alt: 3 },
        obsmaxaz: 0,
        obsmaxel: 1,
        obsmaxrange: 90,
        obsminaz: -100,
        obsminel: -29.45,
        obsminrange: -100,
        shortName: 'http://example.com/showcalendar.html?token=CKF50YzIHxCTKMAg',
        staticNum: 1000,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'http://base.com',
        volume: false,
        zoom: "['39.2233', '-78.8613']",
      },
      {
        static: false,
        linkIridium: true,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: 0,
        obsmaxrange2: 90,
        obsminrange2: 400,
        obsminel2: 4,
        obsmaxaz2: 100,
        obsminaz2: 1,
        alt: 50,
        beamwidth: 24000,
        changeObjectInterval: 0,
        country: 'United States',
        lat: 0,
        linkAehf: false,
        linkWgs: true,
        lon: 45,
        name: 'Jean-Philippe',
        observerGd: { lat: -180, lon: -180, alt: 50 },
        obsmaxaz: -5.48,
        obsmaxel: -5.48,
        obsmaxrange: 400,
        obsminaz: -5.48,
        obsminel: -0.5,
        obsminrange: -5.48,
        shortName: 'http://example.com/showcalendar.html?token=CKF50YzIHxCTKMAg',
        staticNum: 1,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'Www.GooGle.com',
        volume: false,
        zoom: "['41.1994', '-115.9206']",
      },
      {
        static: false,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: true,
        obsmaxel2: 1,
        obsmaxrange2: 50,
        obsminrange2: 550,
        obsminel2: 410,
        obsmaxaz2: 1,
        obsminaz2: 100,
        alt: 5,
        beamwidth: 2048,
        changeObjectInterval: 100,
        country: 'France',
        lat: 45,
        linkAehf: false,
        linkWgs: false,
        lon: -180,
        name: 'Edmond',
        observerGd: { lat: -90, lon: 0, alt: 3 },
        obsmaxaz: -100,
        obsmaxel: -5.48,
        obsmaxrange: 100,
        obsminaz: -5.48,
        obsminel: 0.5,
        obsminrange: 100,
        shortName: 'Www.GooGle.com',
        staticNum: 10,
        sun: '1.0.0',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'Www.GooGle.com',
        volume: true,
        zoom: "['39.2233', '-78.8613']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: -100,
        obsmaxrange2: 70,
        obsminrange2: 100,
        obsminel2: 70,
        obsmaxaz2: -5.48,
        obsminaz2: 1,
        alt: 5,
        beamwidth: 0,
        changeObjectInterval: 100,
        country: 'United States',
        lat: 90,
        linkAehf: true,
        linkWgs: false,
        lon: -45,
        name: 'Edmond',
        observerGd: { lat: 90, lon: -90, alt: 50 },
        obsmaxaz: -5.48,
        obsmaxel: 0,
        obsmaxrange: 350,
        obsminaz: 0,
        obsminel: 1.0,
        obsminrange: -5.48,
        shortName: 'http://example.com/showcalendar.html?token=CKF50YzIHxCTKMAg',
        staticNum: 10,
        sun: 'v4.0.0-rc.4',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'www.google.com',
        volume: false,
        zoom: "['-84.8577', '134.8560']",
      },
      {
        static: true,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: true,
        obsmaxel2: 0,
        obsmaxrange2: 380,
        obsminrange2: 100,
        obsminel2: 520,
        obsmaxaz2: 0,
        obsminaz2: 1,
        alt: 5,
        beamwidth: 200,
        changeObjectInterval: 100,
        country: 'FR',
        lat: 180,
        linkAehf: false,
        linkWgs: false,
        lon: 0,
        name: 'Jean-Philippe',
        observerGd: { lat: 0, lon: -180, alt: 3 },
        obsmaxaz: 0,
        obsmaxel: -100,
        obsmaxrange: 400,
        obsminaz: 1,
        obsminel: 1.0,
        obsminrange: 0,
        shortName: 'https://twitter.com/path?abc',
        staticNum: 1000,
        sun: 'v1.2.4',
        type: SpaceObjectType.SpaceObjectType.ROCKET_BODY,
        url: 'www.google.com',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
    ];
    let result: any = findBestPass.findBestPass(apiMocks.defaultSat, param2);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let param2: any = [
      {
        static: true,
        linkIridium: true,
        linkGalileo: false,
        linkStarlink: false,
        obsmaxel2: 1,
        obsmaxrange2: 50,
        obsminrange2: 100,
        obsminel2: 400,
        obsmaxaz2: -100,
        obsminaz2: 100,
        alt: 50,
        beamwidth: 200,
        changeObjectInterval: 0,
        country: 'China',
        lat: 90,
        linkAehf: true,
        linkWgs: true,
        lon: 45,
        name: 'Edmond',
        observerGd: { lat: -180, lon: -180, alt: 3 },
        obsmaxaz: 100,
        obsmaxel: 100,
        obsmaxrange: 550,
        obsminaz: -100,
        obsminel: 0.5,
        obsminrange: 100,
        shortName: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
        staticNum: 1000,
        sun: '1.0.0',
        type: SpaceObjectType.SpaceObjectType.LAUNCH_SITE,
        url: 'https://',
        volume: true,
        zoom: "['39.2233', '-78.8613']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: true,
        linkStarlink: false,
        obsmaxel2: 100,
        obsmaxrange2: 4,
        obsminrange2: 380,
        obsminel2: 50,
        obsmaxaz2: -5.48,
        obsminaz2: 100,
        alt: 50,
        beamwidth: 24,
        changeObjectInterval: 100,
        country: 'France',
        lat: -90,
        linkAehf: false,
        linkWgs: true,
        lon: -90,
        name: 'Michael',
        observerGd: { lat: -180, lon: 90, alt: 1 },
        obsmaxaz: 0,
        obsmaxel: 0,
        obsmaxrange: 320,
        obsminaz: -100,
        obsminel: 1.0,
        obsminrange: 1,
        shortName: 'https://api.telegram.org/',
        staticNum: 1,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.LAUNCH_SITE,
        url: 'ponicode.com',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
      {
        static: true,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: true,
        obsmaxel2: 100,
        obsmaxrange2: 350,
        obsminrange2: 30,
        obsminel2: 100,
        obsmaxaz2: 1,
        obsminaz2: 1,
        alt: 1,
        beamwidth: 2048,
        changeObjectInterval: 100,
        country: 'France',
        lat: 45,
        linkAehf: false,
        linkWgs: true,
        lon: 90,
        name: 'George',
        observerGd: { lat: -180, lon: -90, alt: 1 },
        obsmaxaz: 0,
        obsmaxel: 0,
        obsmaxrange: 1,
        obsminaz: 1,
        obsminel: 0.5,
        obsminrange: 0,
        shortName: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
        staticNum: 1000,
        sun: '^5.0.0',
        type: SpaceObjectType.SpaceObjectType.LAUNCH_SITE,
        url: 'https://twitter.com/path?abc',
        volume: false,
        zoom: "['-84.8577', '134.8560']",
      },
      {
        static: false,
        linkIridium: true,
        linkGalileo: true,
        linkStarlink: false,
        obsmaxel2: -100,
        obsmaxrange2: 1,
        obsminrange2: 4,
        obsminel2: 4,
        obsmaxaz2: -100,
        obsminaz2: 1,
        alt: 3,
        beamwidth: 9,
        changeObjectInterval: -5.48,
        country: 'US',
        lat: 180,
        linkAehf: true,
        linkWgs: false,
        lon: -180,
        name: 'Anas',
        observerGd: { lat: 180, lon: -180, alt: 3 },
        obsmaxaz: -100,
        obsmaxel: -5.48,
        obsmaxrange: 520,
        obsminaz: -100,
        obsminel: 10.23,
        obsminrange: 0,
        shortName: 'https://accounts.google.com/o/oauth2/revoke?token=%s',
        staticNum: 1,
        sun: '4.0.0-beta1\t',
        type: SpaceObjectType.SpaceObjectType.LAUNCH_SITE,
        url: 'http://base.com',
        volume: true,
        zoom: "['-19.3366', '-46.1477']",
      },
      {
        static: false,
        linkIridium: false,
        linkGalileo: false,
        linkStarlink: true,
        obsmaxel2: 0,
        obsmaxrange2: 70,
        obsminrange2: 520,
        obsminel2: 520,
        obsmaxaz2: 0,
        obsminaz2: 1,
        alt: 3,
        beamwidth: 12,
        changeObjectInterval: -5.48,
        country: 'US',
        lat: 0,
        linkAehf: true,
        linkWgs: false,
        lon: -180,
        name: 'George',
        observerGd: { lat: 180, lon: 180, alt: 5 },
        obsmaxaz: 1,
        obsmaxel: 0,
        obsmaxrange: 30,
        obsminaz: -100,
        obsminel: -1.0,
        obsminrange: -100,
        shortName: 'www.google.com',
        staticNum: 1000,
        sun: 'v1.2.4',
        type: SpaceObjectType.SpaceObjectType.LAUNCH_SITE,
        url: 'https://',
        volume: false,
        zoom: "['61.0865', '-144.7750']",
      },
    ];
    let result: any = findBestPass.findBestPass(apiMocks.defaultSat, param2);
    expect(result).toMatchSnapshot();
  });

  it('should handle a varierty of results 1', () => {
    const result = () => findBestPass.findBestPass(defaultSat, []);
    expect(result).not.toThrow();
  });

  it('should handle a varierty of results 8', () => {
    jest
      .spyOn(checkIsInView, 'checkIsInView')
      .mockImplementationOnce(() => true)
      .mockImplementationOnce(() => false)
      .mockImplementationOnce(() => true)
      .mockImplementationOnce(() => true);
    const result = () => findBestPass.findBestPass(defaultSat, [defaultSensor]);
    expect(result).not.toThrow();
  });
});
