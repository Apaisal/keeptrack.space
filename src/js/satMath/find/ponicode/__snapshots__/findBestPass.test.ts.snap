// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findBestPass.findBestPass 0 1`] = `Array []`;

exports[`findBestPass.findBestPass 1 1`] = `Array []`;

exports[`findBestPass.findBestPass 2 1`] = `Array []`;

exports[`findBestPass.findBestPass 3 1`] = `Array []`;

exports[`findBestPass.findBestPass 4 1`] = `Array []`;
